src/lib.rs

```diff
@@ -1,114 +1,168 @@
 // This check is new and seems buggy (possibly with PyO3 interaction)
 #![allow(clippy::borrow_deref_ref)]
 
-use std::collections::HashSet;
+use std::collections::{BTreeMap, BTreeSet, HashSet};
+use std::iter::successors;
+use std::num::NonZeroU64;
 use std::thread;
 
 use fancy_regex::Regex;
 use pyo3::exceptions;
 use pyo3::prelude::*;
-use pyo3::types::{PyBytes, PyList, PyTuple};
+use pyo3::pyclass;
 use pyo3::PyResult;
+use pyo3::types::{PyBytes, PyList, PyTuple};
 use rustc_hash::FxHashMap as HashMap;
 
-fn _byte_pair_merge<T>(
+type Rank = u32;
+
+const LARGE_ENCODER_CHARACTER_LIMIT: usize = 500;
+
+fn _byte_pair_merge(
+    ranks: &HashMap<Vec<u8>, Rank>,
     piece: &[u8],
-    ranks: &HashMap<Vec<u8>, usize>,
-    f: impl Fn(std::ops::Range<usize>) -> T,
-) -> Vec<T> {
-    // This is a vector of (start, rank).
-    // The rank is of the byte pair starting at position start.
-    // The rank of the last item in the vector is not a valid value.
-    let mut parts: Vec<(usize, usize)> = (0..piece.len() + 1).map(|i| (i, usize::MAX)).collect();
-
-    let get_rank = {
-        #[inline(always)]
-        |parts: &Vec<(usize, usize)>, start_idx: usize, skip: usize| {
-            if (start_idx + skip + 2) < parts.len() {
-                ranks
-                    .get(&piece[parts[start_idx].0..parts[start_idx + skip + 2].0])
-                    .copied()
-            } else {
-                None
-            }
-        }
+) -> Vec<(usize, Rank)> {
+    if piece.len() < LARGE_ENCODER_CHARACTER_LIMIT {
+        _byte_pair_merge_small(ranks, piece) // Quadratic, but lightweight
```

Comment:
quadtratic algo is usually faster for very small words - which is always the case for natural language, but e.g. DNA sequences just a DOS attack